rllib/core/models/torch/base.py

```diff
@@ -0,0 +1,97 @@
+import abc
+from typing import Union
+
+from ray.rllib.core.models.base import Model, ModelConfig
+from ray.rllib.core.models.torch.primitives import nn
+from ray.rllib.models.specs.checker import (
+    is_input_decorated,
+    is_output_decorated,
+    check_input_specs,
+    check_output_specs,
+)
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.utils.typing import TensorType
+
+
+def _raise_not_decorated_exception(input_or_output):
+    raise ValueError(
+        f"`TorchModel.forward()` not decorated with {input_or_output} specification. "
+        f"Decorate it with @check_{input_or_output}_specs() to define a specification."
+    )
+
+
+class TorchModel(nn.Module, Model, abc.ABC):
+    """Base class for RLlib's PyTorch models.
+
+    This class defines the interface for RLlib's PyTorch models and checks
+    whether inputs and outputs of forward are checked with `check_input_specs()` and
+    `check_output_specs()` respectively.
+
+    Example usage for a single Flattening layer:
+
+    .. testcode::
+
+        from ray.rllib.core.models.torch.base import TorchModel
+        from ray.rllib.core.models.base import ModelConfig
+        import torch
+
+        class FlattenModelConfig(ModelConfig):
+            def build(self, framework: str):
+                assert framework == "torch"
+                return TorchFlattenModel(self)
+
+        class TorchFlattenModel(TorchModel):
+            def __init__(self, config):
+                TorchModel.__init__(self, config)
+                self.flatten_layer = torch.nn.Flatten()
+
+            def _forward(self, inputs, **kwargs):
+                return self.flatten_layer(inputs)
+
+        model = FlattenModelConfig().build("torch")
+        inputs = torch.Tensor([[[1, 2]]])
+        print(model(inputs))
+
+    .. testoutput::
+
+        tensor([[1., 2.]])
+
+    """
+
+    def __init__(self, config: ModelConfig, skip_nn_module_init=False):
+        """Initialized a TorchModel.
+
+        Args:
+            config: The ModelConfig to use
+            skip_nn_module_init: Whether to skip the call to nn.Module.__init__.
+                This should be used if the child class has called nn.Module.__init__
+                already to avoid deleting sub-modules that were added between the two
+                nn.Module.__init__ calls.
+        """
+        if not skip_nn_module_init:
```

Comment:
This is weird. sub-classes should always call super().__init__() at the very beginning.

Reply:
Yes. But here are the issues:

- We need to instantiate the wrapped encoder(s) in order to get their specs.
- So we need to do something like `self.encoder = blah` and AFTER that something like `self.input_spec = self.encoder.input_spec()`. 
- The problem is that we need to call nn.Module.__init__() before `self.encoder = blah` because torch enforces it. 
- But that also means that we can not call in in TorchModel anymore, because that would delete `self.encoder`.
- If we were to put TorchModel.__init__() BEFORE all of this, it would call Model.__init__() which in turn already calls `self.get_input_spec()` (which does not work because it needs `self.encoder` in the case of ActorCrtiticEncoder)

Reply:
That's why we have to call get_input_specs() only in `__post_init__()` :) Only after instantiating the whole module you compute the specs that depend on other specs.

Reply:
I have gone through all of these thought process once with RLModules, man :). There is a reason for that `__postinit__()` plugin that I added to the constructor of the base class.

Reply:
😄

Reply:
@kouroshHakha Ok, changed. Python is beatiful.

Reply:
Nice 👍